transaction/std/src/memo_builder/gift_code_sender_memo_builder.rs

```diff
@@ -0,0 +1,276 @@
+    attempted_invalid_note: bool,
+}
+
+// Create an empty GiftCodeSenderMemoBuilder
+impl Default for GiftCodeSenderMemoBuilder {
+    fn default() -> Self {
+        Self {
+            note: "".into(),
+            gift_code_change_memo_enabled: true,
+            wrote_change_memo: false,
+            attempted_invalid_note: false,
+        }
+    }
+}
+
+impl GiftCodeSenderMemoBuilder {
+    /// Set a utf-8 note (up to 64 bytes) onto the sender memo indicating
+    /// any desired info about the gift code. This method will enforce the
+    /// 64 byte limit with a NewMemoErr if the &str passed is longer than
+    /// 64 bytes.
+    pub fn set_gift_code_sender_note(&mut self, note: &str) -> Result<(), NewMemoError> {
+        if note.len() > GiftCodeSenderMemo::MEMO_DATA_LEN {
+            self.attempted_invalid_note = true;
+            return Err(NewMemoError::BadInputs(
+                "Sender note cannot be longer than 64 bytes".into(),
+            ));
+        }
+        self.note = note.into();
+        self.attempted_invalid_note = false;
+        Ok(())
+    }
+    /// Clear the gift code sender note
+    pub fn clear_sender_note(&mut self) {
+        self.attempted_invalid_note = false;
+        self.note = "".into();
+    }
+    /// Enable change memos
+    pub fn enable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = true;
+    }
+    /// Disable change memos
+    pub fn disable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = false;
+    }
+}
+
+impl MemoBuilder for GiftCodeSenderMemoBuilder {
+    /// Set the fee
+    fn set_fee(&mut self, _fee: Amount) -> Result<(), NewMemoError> {
+        Ok(())
+    }
+
+    /// Gift code destination memos are not allowed - all gift code
+    /// memos accompany TxOuts sent to the change address
+    fn make_memo_for_output(
+        &mut self,
+        _amount: Amount,
+        _recipient: &PublicAddress,
+        _memo_context: MemoContext,
+    ) -> Result<MemoPayload, NewMemoError> {
+        Err(NewMemoError::DestinationMemoNotAllowed)
+    }
+
+    /// Build a memo for a gift code change output
+    fn make_memo_for_change_output(
+        &mut self,
+        _amount: Amount,
+        _change_destination: &ReservedDestination,
+        _memo_context: MemoContext,
+    ) -> Result<MemoPayload, NewMemoError> {
+        if !self.gift_code_change_memo_enabled {
+            return Ok(UnusedMemo {}.into());
+        };
+        if self.wrote_change_memo {
+            return Err(NewMemoError::MultipleChangeOutputs);
+        };
+        // Prevent callers from writing memo if last note set attempt was a failure
+        if self.attempted_invalid_note {
+            return Err(NewMemoError::BadInputs(
+                "Tried to set a note longer than 64 bytes".into(),
+            ));
+        }
+        self.wrote_change_memo = true;
+        Ok(GiftCodeSenderMemo::new(self.note.as_str())?.into())
```

Comment:
nit: 
```suggestion
        Ok(GiftCodeSenderMemo::new(&self.note)?.into())
```

Reply:
I think I'm going to keep it this way to express what's happening. The & is concise but hides more of what's happening.